### PR TITLE
Bucket lifecycle rule creation policies compatibility with aws new API

### PR DESCRIPTION
**Pull request by liranmauda**
### Explain the changes

***This PR intention is not to add any flow if there are missing, It will change the API to be compatible with AWS's new API.
due to the fact that we are not testing it, if there are missing flows they should be addressed on a different PR***

- Fixing bucket lifecycle rule creation policies to be compatible with AWS new API
- Adding Upgrade script

https://docs.aws.amazon.com/AmazonS3/latest/API/API_LifecycleRule.html

Signed-off-by: liranmauda <liran.mauda@gmail.com>

### Issues: Fixed #xxx / Gap #xxx
1.  https://bugzilla.redhat.com/show_bug.cgi?id=2029298

### CODE REVIEW

File: src/upgrade/upgrade_scripts/5.10.0/upgrade_bucket_lifecycle_structure.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            if (!bucket.lifecycle_configuration_rules || bucket.deleting) return;
+
+            let iteratee = 0;
+            for (const lifecycle_rule of bucket.lifecycle_configuration_rules) {
+                if (lifecycle_rule.prefix) {
+                    dbg.log0('bucket.lifecycle_configuration_rules[', iteratee, ']', lifecycle_rule);
+                    lifecycle_rule.filter = {};
+                    lifecycle_rule.filter.prefix = lifecycle_rule.prefix;
+                    delete lifecycle_rule.prefix;
+                    bucket.lifecycle_configuration_rules[iteratee] = lifecycle_rule;
```
Comment (jackyalbo) [inline src/upgrade/upgrade_scripts/5.10.0/upgrade_bucket_lifecycle_structure.js:23]: I'm not sure why you need to do it inline - I think it's better to just create a new array of changes and push when needed, without changing the system_store (at least not before the make_changes)

File: src/endpoint/s3/ops/s3_put_bucket_lifecycle.js [hunk 2]
```diff
@@ -13,30 +13,42 @@ async function put_bucket_lifecycle(req) {
...
+        const current_rule = {
             id: rule_id,
-            prefix: rule.Prefix[0],
+            filter: {},
             status: rule.Status[0]
         };
+        if (rule.Prefix) {
+            throw new S3Error(S3Error.InvalidArgument);
+        }
+        if (rule.Filter) {
```
Comment (jackyalbo) [inline src/endpoint/s3/ops/s3_put_bucket_lifecycle.js:37]: what about all the other options in filter?  If we don't want to handle maybe we should throw something...

File: src/api/common_api.js [hunk 3]
```diff
@@ -848,5 +848,51 @@ module.exports = {
                 }
             }
         },
+        bucket_lifecycle_filter: {
+            type: 'object',
+            properties: {
+                and: {
+                    type: 'object',
```
Comment (jackyalbo) [inline src/api/common_api.js:855]: not sure what it this and? It seems like the filter needs to be an array of this properties. take a look here: https://docs.aws.amazon.com/AmazonS3/latest/userguide/lifecycle-configuration-examples.html

**Review by jackyalbo — CHANGES_REQUESTED** [review]

File: src/endpoint/s3/ops/s3_put_bucket_lifecycle.js (hunk 2 above)
Comment (liranmauda) [reply src/endpoint/s3/ops/s3_put_bucket_lifecycle.js:37]: Sure, We will throw `S3Error.NotImplemented`

File: src/api/common_api.js (hunk 3 above)
Comment (liranmauda) [reply src/api/common_api.js]: As we talked, removed the `and`, and we will just do it when we parse the request.
also as this PR is not containing new/missing flows commented out the unused properties

File: src/api/bucket_api.js [hunk 4]
```diff
@@ -614,17 +614,17 @@ module.exports = {
                         type: 'array',
                         items: {
                             type: 'object',
-                            required: ['id', 'prefix', 'status', 'expiration'],
+                            required: ['id', 'status'],
```
Comment (dannyzaken) [inline src/api/bucket_api.js:617]: isn't expiration required?

File: src/endpoint/s3/ops/s3_put_bucket_lifecycle.js [hunk 5]
```diff
@@ -13,30 +13,45 @@ async function put_bucket_lifecycle(req) {
...
+    //     <Prefix>logs/</Prefix> 
+    //   </Filter>
     //   <Status>Enabled</Status>
     //   <Expiration>
     //     <Days>365</Days>
     //   </Expiration>
     // </Rule>
     const lifecycle_rules = _.map(req.body.LifecycleConfiguration.Rule, rule => {
-        var rule_id = uuid().split('-')[0];
+        let rule_id = uuid().split('-')[0];
```
Comment (dannyzaken) [inline src/endpoint/s3/ops/s3_put_bucket_lifecycle.js:25]: why do you split the uuid?

File: src/upgrade/upgrade_scripts/5.10.0/upgrade_bucket_lifecycle_structure.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+async function run({ dbg, system_store, system_server }) {
+
+    try {
+        dbg.log0('starting upgrade lifecycle...');
+        let should_make_change = false;
+        const buckets = [];
+        for (const bucket of system_store.data.buckets) {
+            //Do not update if there are no lifecycle_configuration_rules or if the bucket is being deleted.
+            if (!bucket.lifecycle_configuration_rules || bucket.deleting) return;
```
Comment (dannyzaken) [inline src/upgrade/upgrade_scripts/5.10.0/upgrade_bucket_lifecycle_structure.js:14]: why do you ignore deleting? I think we want to change the schema anyway.

File: src/upgrade/upgrade_scripts/5.10.0/upgrade_bucket_lifecycle_structure.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+                expiration: lifecycle_configuration_rules.expiration,
+            }));
+
+            buckets.push({
+                _id: bucket._id,
+                lifecycle_configuration_rules: new_rules,
+            });
+        }
+
+        if (new_rules.length > 0) {
```
Comment (jackyalbo) [inline src/upgrade/upgrade_scripts/5.10.0/upgrade_bucket_lifecycle_structure.js:30]: Undefined new_rules will break your code. Check buckets length instead and then move new_rules declaration to inside the for.

**Review by jackyalbo — CHANGES_REQUESTED** [review]

**Review by jackyalbo — APPROVED** [review]

**Pull request merged**